homeassistant/components/input_select/__init__.py

```diff
@@ -183,138 +185,130 @@ class InputSelectStorageCollection(collection.StorageCollection):
+        if not state or state.state not in self.options:
+            self._attr_current_option = self.options[0]
         else:
-            self._current_option = state.state
-
-    @property
-    def should_poll(self):
-        """If entity should be polled."""
-        return False
-
-    @property
-    def name(self):
-        """Return the name of the select input."""
-        return self._config.get(CONF_NAME)
-
-    @property
-    def icon(self):
-        """Return the icon to be used for this entity."""
-        return self._config.get(CONF_ICON)
-
-    @property
-    def _options(self) -> list[str]:
-        """Return a list of selection options."""
-        return self._config[CONF_OPTIONS]
-
-    @property
-    def state(self):
-        """Return the state of the component."""
-        return self._current_option
+            self._attr_current_option = state.state
 
     @property
-    def extra_state_attributes(self):
+    def extra_state_attributes(self) -> dict[str, bool]:
         """Return the state attributes."""
-        return {ATTR_OPTIONS: self._config[ATTR_OPTIONS], ATTR_EDITABLE: self.editable}
+        return {ATTR_EDITABLE: self.editable}
 
-    @property
-    def unique_id(self) -> str | None:
-        """Return unique id for the entity."""
-        return self._config[CONF_ID]
-
-    @callback
-    def async_select_option(self, option):
+    async def async_select_option(self, option: str) -> None:
         """Select new option."""
-        if option not in self._options:
+        if option not in self.options:
             _LOGGER.warning(
                 "Invalid option: %s (possible options: %s)",
                 option,
-                ", ".join(self._options),
+                ", ".join(self.options),
             )
             return
-        self._current_option = option
+        self._attr_current_option = option
         self.async_write_ha_state()
 
     @callback
-    def async_select_index(self, idx):
+    def async_select_index(self, idx: int) -> None:
         """Select new option by index."""
-        new_index = idx % len(self._options)
-        self._current_option = self._options[new_index]
+        new_index = idx % len(self.options)
+        self._attr_current_option = self.options[new_index]
         self.async_write_ha_state()
 
     @callback
-    def async_offset_index(self, offset, cycle):
+    def async_offset_index(self, offset: int, cycle: bool) -> None:
         """Offset current index."""
-        current_index = self._options.index(self._current_option)
+        if self.current_option is None:
+            self._attr_current_option = self.options[0]
+
+        current_index = (
+            self.options.index(self.current_option)
+            if self.current_option is not None
+            else 0
+        )
         new_index = current_index + offset
         if cycle:
-            new_index = new_index % len(self._options)
+            new_index = new_index % len(self.options)
```

Comment:
Doesn't this result in the second option (i.e., `new_index = 1`) being selected when current option is `None` or not in the list? I think you want to bypass offset entirely in that case.

Reply:
Not it does not, this only takes care of rollovers.

> or not in the list

That is not a possible situation.

Reply:
When `self.current_option` is `None`, `current_index` becomes `0`. Then you call `new_index = current_index + offset` and end up with 1 (or -1)

Reply:
Yup, seems good to me.

PS: This is existing code.

Reply:
This line isn't correct though:

> otherwise it will reset to the first option from the newly provided list